pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,430 @@
+		latestFeatureStates: make([]featurestatesv1alpha1.FeatureState, 0),
+	}
+	var err error
+	// This is idempotent if CRD is pre-created then we continue with initialization of svFSSReplicationService
+	err = k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
+		reflect.TypeOf(featurestatesv1alpha1.CnsCsiSvFeatureStates{}).Name(), CRDGroupName, internalapis.SchemeGroupVersion.Version, apiextensionsv1beta1.NamespaceScoped)
+	if err != nil {
+		log.Errorf("failed to create CnsCsiSvFeatureStates CRD. Error: %v", err)
+		return err
+	}
+
+	// Create the kubernetes client
+	k8sClient, err = k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("create k8s client failed. Err: %v", err)
+		return err
+	}
+	// get kube config
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Error: %v", err)
+		return err
+	}
+	// create controller runtime client
+	controllerRuntimeClient, err = k8s.NewClientForGroup(ctx, config, CRDGroupName)
+	if err != nil {
+		log.Errorf("failed to create controllerRuntimeClient. Err: %v", err)
+		return err
+	}
+
+	pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
+	if err != nil {
+		log.Errorf("failed to retrieve latest feature switch states. Error: %v", err)
+		return err
+	}
+
+	// Create k8s Informer and watch on configmaps and namespaces
+	informer := k8s.NewInformer(k8sClient)
+	// configmap informer to watch on SV featurestate config-map
+	informer.AddConfigMapListener(ctx, k8sClient, svFeatureStateConfigMapNamespace,
+		// Add
+		func(Obj interface{}) {
+			configMapAdded(Obj)
+		},
+		// Update
+		func(oldObj interface{}, newObj interface{}) {
+			configMapUpdated(oldObj, newObj)
+		},
+		// Delete
+		func(obj interface{}) {
+			configMapDeleted(obj)
+		})
+
+	// namespace informer to watch on namespaces
+	informer.AddNamespaceListener(
+		// Add
+		func(obj interface{}) {
+			namespaceAdded(obj)
+		},
+		// Update
+		func(oldObj interface{}, newObj interface{}) {
+			namespaceUpdated(oldObj, newObj)
+		},
+		// Delete
+		func(obj interface{}) {
+			namespaceDeleted(obj)
+		})
+	informer.Listen()
+	log.Infof("informer on config-map and namespaces started")
+
+	// Start routine to process pending feature state updates
+	go pendingCRUpdatesObj.processPendingCRUpdates()
+	log.Infof("started background routine to process pending feature state updates at regular interval")
+	log.Infof("SvFSSReplicationService is running")
+	var stopCh = make(chan bool)
+	<-stopCh
+	return nil
+}
+
+// processPendingCRUpdates helps process pending CR updates at regular interval
+func (pendingCRUpdatesObj *pendingCRUpdates) processPendingCRUpdates() {
+	ticker := time.NewTicker(crUpdateRetryInterval)
+	for range ticker.C {
+		func() {
+			pendingCRUpdatesObj.lock.Lock()
+			defer pendingCRUpdatesObj.lock.Unlock()
+
+			ctx, log := logger.GetNewContextWithLogger()
+			if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
```

Comment:
Does this mean, we may have stale replicated FSS config maps in all supervisor namespaces? If yes, how can we solve this? Since we do not expect anyone to delete the supervisor FSS config map, we should be fine. I'm worried about a code change that deletes the supervisor config map in the future. The current code change is silently ignoring the problem. I would suggest to crash the container so that we catch such problems early during dev. Basically if there is no supervisor FSS or if it is deleted during runtime, we crash the CSI container and CSI will never come up. This is a nice way to get informed about such issues early. Thoughts?

Reply:
Addressed this with https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/764#discussion_r606515866